### ISSUE TITLE
Causes bootloop (Pixel 8)

### ISSUE THREAD

**Issue by ostrolucky** [pwnlogs/cert-fixer#2]
Here are the boot error logs. I also tried to run cert-fixer standalone, without any other magisk module enabled.

[boot.log](https://github.com/user-attachments/files/17374574/boot.log)

**Comment by pwnlogs** [comment]
@ostrolucky , Cert-Fixer was developed as a Magisk module. I'm not sure how you managed to run it without Magisk. The script is supposed to run right after `/data` is decrypted. I'm curious about how you run it.

**Comment by ostrolucky** [comment]
I'm using Magisk, I am just saying that even when I also disabled every other Magisk module outside cert-fixer - still causes bootloop

**Comment by pwnlogs** [comment]
@ostrolucky , I went through your log file. These are system logs. Most of these errors are related to memory or hardware, not cert-fixer. Your first error `NetlinkListener: recvmsg failed (No buffer space available)` says that your kernel’s network buffer has run out of memory. May be try booting into recover mode and increase the buffer size?
```
adb shell "echo 262144 > /proc/sys/net/core/rmem_max"
adb shell "echo 262144 > /proc/sys/net/core/wmem_max"
```
Are you using a physical device or emulator? If you are using an emulator, I suggest following [this blog](https://blog.pwnlogs.dev/articles/cert-fixer/index.html) step by step.

**Comment by ostrolucky** [comment]
It's a real physical Pixel 8 device. Logs I provided is complete boot log filtered only by error severity, so yes it contains irrelevant things as well. Not sure how would network buffer related errors be related, so let's leave that at the later possibly, I'm kinda worried about doing such actions to my daily device.

What about those Zygote errors in combination with certificates (`ExceptionInInitializerError` coming from `SystemCertificateSource.java`? Those could be related. Also `com.android.org.conscrypt.OpenSSLX509CertificateFactory$ParsingException`, I can see in zip file you are shipping there is a `post-fs-data.sh` file containing references to `com.android.conscrypt` stuff.

**Comment by pwnlogs** [comment]
@ostrolucky , yeah that's right. Are you able to get `/data/local/tmp/cert-fixer.log` from the device? This would have Cert-Fixer logs. We would know if the execution didn't complete as expected.

Like you said I see this in your logs:
```
FileIntegrityService: Invalid certificate, ignored: java.security.cert.CertificateException: com.android.org.conscrypt.OpenSSLX509CertificateFactory$ParsingException: com.android.org.conscrypt.OpenSSLX509CertificateFactory$ParsingException: java.lang.RuntimeException: error:0c00006d:ASN.1 encoding routines:OPENSSL_internal:DECODE_ERROR
```
But I will be surprised if there is a certificate parsing error for a certificate that was copied by Cert-Fixer. Cert-Fixer copies the certificates from the user cert store to the system store. Expectation is that if the user store certificate import was successful, the certificate has the correct format, and copying it to the system store will be safe.

`/data/local/tmp/cert-fixer.certs-found` will have a list of certificates Cert-Fixer found in your user store. Can you try checking if these certs are in the right format?

**Comment by ostrolucky** [comment]
[cert-fixer.log](https://github.com/user-attachments/files/17529528/cert-fixer.log)

`cert-fixer.certs-found` contains these certificates
```
145b938d
4f74014f
```

That's these certificates [cacerts-added.zip](https://github.com/user-attachments/files/17529610/cacerts-added.zip)

**Comment by pwnlogs** [comment]
Thanks @ostrolucky , I see
`cp: can't stat '/apex/com.android.conscrypt/cacerts/*': No such file or directory` error at [this line](https://github.com/pwnlogs/cert-fixer/blob/1b8e3fe192e4e682871e1a786da0dcc4e5fc8b7e/Cert-Fixer/post-fs-data.sh#L69C15-L69C52). That's surprising, because the [`if`](https://github.com/pwnlogs/cert-fixer/blob/1b8e3fe192e4e682871e1a786da0dcc4e5fc8b7e/Cert-Fixer/post-fs-data.sh#L61) condition exists to check the existance of this directory and avoid this situation.

Also there is an error (`rmdir: '/data/local/tmp/adg-ca-copy': Device or resource busy`) at [the line](https://github.com/pwnlogs/cert-fixer/blob/1b8e3fe192e4e682871e1a786da0dcc4e5fc8b7e/Cert-Fixer/post-fs-data.sh#L86) trying to remove the temporary directory.  I suspect the unmount was not successful or a process is still using this directory. We could choose to not remove the directory.

I don't have access to a device today, I will test further in a day or two. Sorry if this caused the boot loop.

**Comment by ostrolucky** [comment]
Is it possible that script runs multiple times in parallel for some reason and that causes race conditions?

**Comment by pwnlogs** [comment]
The script is run by Magisk. It triggers the `post-fs-data.sh` script right after `/data` is decrypted. I'm not aware of any bug in Magisk that could cause the script to run twice.

The mount operation is not persistent across boot. So, not removing the directory is something to consider. But I would first try it in an emulator.

**Comment by bdr99** [comment]
I also experienced this bootloop on a Pixel 6a with only this module installed. It was working fine at first. Then I installed a new cert to the user cert store, and when I rebooted after that, that's when the bootloop happened.

**Comment by pwnlogs** [comment]
@bdr99 and @ostrolucky , I haven't been able to reproduce this issue. Please let me know if you manages to find the root cause.

**Comment by ostrolucky** [comment]
Have you tried with the same certificates I posted? Could be certificate specific

**Comment by pwnlogs** [comment]
@ostrolucky , My sincere appologies, I should have tried this before. Your `HTTP Toolkit CA` cert is causing the issue. That was totally unexpected. The idea of first installing the certificate in the user store was to avoid any issues with the certificate. But apparantly, that was not sufficient. I'm still trying to figure out what is wrong with the certificate.

@ostrolucky  and @bdr99 ,
To stop your boot loop, you can edit the Magisk module so that it removes the problamatic certificate. You can do this while the deive is stuck in boot loop. While the device is in the boot loop, get adb root shell and edit  the file: `/data/adb/modules/cert-fixer/post-fs-data.sh`. Add the following line before the comment `# module root` and reboot.
```
find / -name 4f74014f.0 2>/dev/null | xargs rm
# where 4f74014f.0 is your certificate that is causing the issue.
```
That should fix the boot loop. 

If I manage to find what is wrong with the certificate, I will let you know. @bdr99 , if you could please send me your certificate, it will be very helpful in identifying the issue.

**Comment by bdr99** [comment]
I did some testing and was able to learn something. I think the cause of the bootloop is not related to the specific certificate, but rather to just having multiple certificates installed in the user cert store at the same time.

These are the two certificates I'm using:
[certfixer_issue.zip](https://github.com/user-attachments/files/17983851/certfixer_issue.zip)

I followed these steps:
1. Did a clean install of the latest android 14 factory image on a Pixel 6a
2. Installed Magisk 28.0 and rebooted
3. Installed cert-fixer 1.0 and rebooted
4. Installed the Fiddler cert into the user cert store and rebooted
5. Confirmed the cert-fixer module is still active in Magisk and the Fiddler cert shows in the system trust store
6. Installed the Charles cert into the user cert store and rebooted
7. A boot loop occurs, triggering Magisk's safe mode, which disables the cert-fixer module

Other observations:
- The order of installing the two certs doesn't matter, the bootloop happens regardless of the order.
- If I re-enable the cert-fixer module in Magisk and reboot, it boot loops again.
- If I remove either one of the two certs from the user store, it fixes the boot loop and cert-fixer works correctly.
  - It needs to be deleted from Encryption & credentials > Trusted credentials > User. Deleting it from Encryption & credentials > User credentials doesn't work.

Hope this helps!

**Comment by pwnlogs** [comment]
@bdr99 ,That makes a lot more sense. Thank you. If you have already solved this issue, please create a pull request. Otherwise, I will mention the issue in the readme, until I have time to look at this again.

**Comment by pwnlogs** [comment]
This issue is now fixed in version 1.1. Tested by adding up to 10 certificates simultaneously.
![Screenshot](https://github.com/user-attachments/assets/fafb2f1e-455c-4426-8a6b-9a7e6ba610f0)

**Comment by bdr99** [comment]
@pwnlogs Thank you for fixing this!
